connections/simopen.md

```diff
@@ -0,0 +1,89 @@
+# Simultaneous Open for bootstrapping connections in multistream-select
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active | DRAFT           |
+
+Authors: [@vyzo]
+
+Interest Group: [@raulk], [@stebalien]
+
+[@vyzo]: https://github.com/vyzo
+[@raulk]: https://github.com/raulk
+[@stebalien]: https://github.com/stebalien
+
+See the [lifecycle document][lifecycle-spec] for context about maturity level
+and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+
+## Introduction
+
+In order to support direct connections through NATs with hole
+punching, we need to account for simultaneous open. In such cases,
+there is no single initiator and responder, but instead both peers act
+as initiators. This breaks protocol negotiation in
+multistream-select, which assumes a single initator.
+
+This draft proposes a simple extension to the multistream protocol
+negotiation in order to select a single initator when both peers are
+acting as such.
+
+## The Protocol
+
+When a peer acting as the initiator enters protocol negotiation, it
+sends the string `iamclient` as first protocol selector. If the other
+peers is a responder or doesn't support the extension, then it
+responds with `na` and protocol negotiation continues as normal.
```

Comment:
Note that the [rust-libp2p multistream-select implementation currently requires protocols to start with a `/`](https://github.com/libp2p/rust-libp2p/blob/40c4287bc9a95461889e10612519960a16a2874f/misc/multistream-select/src/protocol.rs#L73-L82) and thus the `iamclient` message is not backwards compatible with rust-libp2p as it would not parse as a valid protocol identifier.

The specification recommends protocols to start with a `/`, but does not require it. The rust-libp2p implementation is thus not in-line with the specification.

> Each protocol supported by a peer is identified using a unique string called a protocol id. While any string can be used, the conventional format is a path-like structure containing a short name and a version number, separated by / characters. For example: /mplex/1.0.0 identifies version 1.0.0 of the mplex stream multiplexing protocol. multistream-select itself has a protocol id of /multistream/1.0.0.

https://github.com/libp2p/specs/tree/master/connections#protocol-negotiation

I am posting this not to ask the specification to change, but only to raise awareness in case other implementations are as strict as rust-libp2p. (//CC @vasco-santos for the js-libp2p side).

Reply:
> Note that the rust-libp2p multistream-select implementation currently requires protocols to start with a /

This was done because the code that parses multistream-select payloads is context-free.
In other words, if we want to parse the string `"na"`, the code that does so doesn't know whether it should interpret it as a "not available" response or as a protocol named `"na"`.

Reply:
It's still reasonable to change this, it's not used in practice anyways.

Really, it would likely make more sense as `/libp2p/simultaneous-connect`.

Reply:
Though initially not intentioned, I am in favor of replacing `iamclient` with `/libp2p/simultaneous-connect`. This would both comply with the above mentioned convention to start (fake) protocol IDs with a `/` and be interoperable with the stateless parser in rust-libp2p.

I have opened https://github.com/libp2p/specs/pull/321 which would do the change in the specification. Please let me know what you think either on https://github.com/libp2p/specs/pull/321 or here.